compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/SceneState.kt

```diff
@@ -0,0 +1,411 @@
+            configuration = configuration,
+            keyboardOverlapHeightState = keyboardOverlapHeightState,
+            viewProvider = { rootViewController.view },
+            sceneStates = sceneStates,
+            densityProvider = densityProvider,
+            sceneStateProvider = { this },
+        )
+    }
+
+    private val keyboardEventHandler: KeyboardEventHandler by lazy {
+        object : KeyboardEventHandler {
+            override fun onKeyboardEvent(event: SkikoKeyboardEvent) {
+                val composeEvent = KeyEvent(event)
+                if (!uiKitTextInputService.onPreviewKeyEvent(composeEvent)) {
+                    scene.sendKeyEvent(composeEvent)
+                }
+            }
+        }
+    }
+
+    private val uiKitTextInputService: UIKitTextInputService by lazy {
+        UIKitTextInputService(
+            updateView = {
+                sceneView.setNeedsDisplay() // redraw on next frame
+                CATransaction.flush() // clear all animations
+            },
+            rootViewProvider = { rootViewController.view },
+            densityProvider = densityProvider,
+            focusStack = focusStack,
+            keyboardEventHandler = keyboardEventHandler
+        )
+    }
+
+    private val delegate: SkikoUIViewDelegate by lazy {
+        SkikoUIViewDelegateImpl(
+            { scene },
+            interopContext,
+            densityProvider,
+        )
+    }
+
+    override fun setContentWithCompositionLocals(content: @Composable () -> Unit) {
+        scene.setContent {
+            /**
+             * TODO isReadyToShowContent it is workaround we need to fix.
+             *  https://github.com/JetBrains/compose-multiplatform-core/pull/861
+             *  Density problem already was fixed.
+             *  But there are still problem with safeArea.
+             *  Elijah founded possible solution:
+             *   https://developer.apple.com/documentation/uikit/uiviewcontroller/4195485-viewisappearing
+             *   It is public for iOS 17 and hope back ported for iOS 13 as well (but we need to check)
+             */
+            if (sceneView.isReadyToShowContent.value) {
+                EntrypointCompositionLocals {
+                    SceneCompositionLocal {
+                        content()
+                    }
+                }
+            }
+        }
+    }
+
+    private val safeAreaState: MutableState<PlatformInsets> by lazy {
+        //TODO It calc 0,0,0,0 on initialization
+        mutableStateOf(calcSafeArea())
+    }
+    private val layoutMarginsState: MutableState<PlatformInsets> by lazy {
+        //TODO It calc 0,0,0,0 on initialization
+        mutableStateOf(calcLayoutMargin())
+    }
+
+    override fun updateSafeArea() {
+        safeAreaState.value = calcSafeArea()
+        layoutMarginsState.value = calcLayoutMargin()
+    }
+
+    @Composable
+    private fun SceneCompositionLocal(content: @Composable () -> Unit) =
```

Comment:
It's still not about scene, but platform specifics

Reply:
I don't think so. 
For now it is better to use Scene terminology.